README
Last modified: 30 Oct 2000

This is Pickle, a C++ library for conveniently calling Perl functions
and examining Perl data.  It provides a higher-level interface than
Perl's native C support, shielding the programmer from complex details
and sacrificing some power and speed for ease of use and maintenance.
It is well suited for large applications prototyped in Perl and
transitioned piece by piece to C++, since it allows one to move the
boundary between implementation languages quickly and easily.

The name `Pickle' is a simplification of `perlcall', as in `perldoc
perlcall'.

The library is documented in the file pickle.pod and should be
browsable via `perldoc pickle' after installation.  Newer versions
will be at http://search.cpan.org/search?dist=pickle and in CVS branch
`pickle-0_5-maint' at http://sourceforge.net/cvs/?group_id=9183.
Pickle has been known to work with Perl v5.6.0, 5.005_03, Linux, and
Solaris (GNU CC), using both default and ithreads (-Dusethreads)
configurations.

This library will *NOT* work with Perl 5.005x and earlier.

Features include:

    * automatic memory management using Perl's reference counts

    * two-way exception handling, Perl->C++ and C++->Perl

    * C++ functions callable from Perl and vice versa

    * read/write access to symbol table, array, and hash elements

    * scalar conversion operators for C++ builtin types

    * application binary compatibility across Perl implementations
      and configurations

This is ALPHA software.  It is incomplete and very likely to contain
bugs.

To build Pickle, you need the Perl build environment and a C++
compiler.  If you used an ordinary C compiler to compile Perl, specify
the C++ compiler as CC and (if necessary) LD.  For example, if you
plan to use g++, try this:

    perl Makefile.PL CC=g++ LD=g++

`make install' puts pickle.hh into $(PREFIX)/include and the following
files and symlinks into $(PREFIX)/lib:

    libperlint.so -> libperlint.so.1
    libpickle.so -> libpickle.so.1

$(PREFIX) here means Perl's base installation directory, typically
/usr/local.  Pickle's Makefile.PL supports only shared library builds,
but there should be no obstacle to putting the object files into a
static library if so desired.

libperlint.so contains a copy of the Perl interpreter.  You may have
to build Perl with special compiler flags to allow it to be made into
a shared library.  For GCC, this means adding `-Accflags=-fPIC' to the
Configure command line.  You need -lperlint when linking Pickle with a
C++ program, but you do not need it for using C++ from within Perl.

libpickle.so is the Pickle library.  -lpickle is always required when
using Pickle.  You have to make arrangements for the libraries to be
found at run time.  See your system's shared library documentation.

Here is the synopsis from pickle.pod:

    #include <pickle.hh>
    using namespace Pickle;

    Interpreter::vivify ();

    eval_string ("use Cwd;");
    Scalar cwd = call_function ("cwd");
    cout << "cwd is " << string (cwd) << endl;

    eval_string ("use File::Glob ':glob';");
    List args = List () << "~/*.doc"
                        << call_function ("GLOB_TILDE");
    Arrayref files = call_function ("glob", args, LIST);
    for (size_t i = 0; i < files.size(); i++)
        cout << string(files[i]) << endl;

    Scalar doit (Scalar& n)
    {
        return double (n) * double (n);
    }
    define_sub ("main", "square", doit);
    cout << int (eval_string ("square(42)")) << endl;

    try { call_some_perl() }
    catch (Exception* e) { cerr << e->what(); delete e; }
    throw new Exception ("Your fault");


Copyright (C) 2000 John Tobey
jtobey@john-edwin-tobey.org  All rights reserved.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; see the file COPYING.  If not, write to the
   Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
   MA 02111-1307  USA

If included in the official Perl distribution at
http://www.perl.com/CPAN/src/, this software may be distributed under
the same terms as said official Perl distribution.  For other
licensing arrangements, contact the author.


ChangeLog
2007-05-23  John Tobey  <jtobey@canuck>

	* Release 0.5.2. Fixes for recent C++ compilers and Perl.

2000-10-30  John Tobey  <jtobey@john-edwin-tobey.org>

	* (0.5 branch): Release 0.5.1.

	* (0.5 branch): Support Perl 5.005.

	* (0.5 branch) pickle.hh, scalar.cc: Scalar::bless - new method

	* (0.5 branch) scalarref.cc: Added Scalarref default constructor.
	Fixed a refcount bug in fetch().

2000-07-23  John Tobey  <jtobey@john-edwin-tobey.org>

	* Release 0.5.  Substantial portions rewritten.  Threaded Perl and
	MULTIPLICITY are now supported.  Documentation improved.

2000-07-12  John Tobey  <jtobey@john-edwin-tobey.org>

	* pickle.hh, interpreter.cc, scalar.cc: replaced
	List::get_arrayref with operators Arrayref& and const Arrayref&.

	* pickle.pod: added XS example.

2000-06-27  John Tobey  <jtobey@john-edwin-tobey.org>

	* Release 0.3.  What happened to 0.2?  Well, the major changes are
	the renaming of the library, easier configuration, and the removal
	of some experimental interfaces.  Minutiae are in CVS log entries.

2000-03-28  John Tobey  <jtobey@john-edwin-tobey.org>

	* Release 0.1.


test_pickle.cc
#include <iostream>
#include "math.h"
#include "pickle.hh"

using namespace Pickle;
using namespace std;

static Interpreter* p;

int
main (int argc, char** real_argv)
{

  try
    {
      //const char* argv[] = { "perl", "-we0" };
      //p = new Interpreter (sizeof argv / sizeof argv [0], argv);
      p = Interpreter::vivify ();
    }
  catch (Init_Exception* e)
    {
      cerr << e->what () << endl;
      exit (1);
    }

  try
    {
      void test_except ();
      test_except ();

      void test_call ();
      test_call ();

      void test_cb ();
      test_cb ();

      Scalar a1[] = { 1, "2.1" };
      Arrayref a (sizeof a1 / sizeof a1 [0], a1);
      cerr << "a has " << a .size () << " elements." << endl;
      a = a;
      cerr << "a is a " << a .ref () .as_string () << endl;

      cerr << "$^X: " << p ->Scalarref ("\030") .fetch () .as_string ()
	   << endl;
      cerr << "$0: " << (string) Scalarref ("0") .fetch () << endl;

      Hashref h;
      h .store ("foo", 5L);
      h .store ("bar", "baz");
      cerr << int (h .fetch ("foo")) << string (h .fetch ("bar")) << endl;

      h .store ("ary", a);
      Arrayref (h .fetch ("ary")) .store (2, 4.9);

      cerr << 2.4 * cos (a[2]) << endl;
      cerr << ((int) a [0] + (int) a [1] + (int) a [2]) << endl;
      Scalar s = "+5.0";
      cerr << s;
      cerr << int (s) << endl;
      cerr << s;

      // Turn on warnings.
      p ->Scalarref ("\027") .store (1);

      // Trigger "use of uninitialized" warning under `-w'.
      cerr << Scalar () .as_int () << endl;

      //Coderef sum = p->compile_sub ("my $x = 0; $x += $_ for @_; $x");

      //cerr << sum .call (List () << 7 << 14.0 << "21") .as_int () <<endl;

    }
  catch (Exception* e)
    {
      cerr << "test: " << e->what () << endl;
      delete e;
      exit (1);
    }
  try
    {
      if (p != 0)
	delete p;
    }
  catch (Exception* e)
    {
      cerr << "test: " << e->what () << endl;
      delete e;
      exit (1);
    }
  cerr << "ping: " << Interpreter::ping () << endl;
  return 0;
}

void
test_except ()
{
  try
    {
      eval_string ("$x = 0; 4 / $x");
    }
  catch (Exception* e)
    {
      cerr << "Exception caught: " << e->what () << endl;
      delete e;
    }
}

void
test_call ()
{
  eval_string ("use Cwd;");
  Scalar cwd = call_function ("cwd");
  cerr << "cwd is " << cwd .as_string () << endl;

  if (Scalarref ("]") .fetch () .as_double () < 5.006)
    {
      cerr << "Skipping File::Glob tests for pre-5.6 Perl.  $] == "
	   << Scalarref ("]") .fetch () .as_double () << endl;
    }
  else
    {
      eval_string ("use File::Glob ':glob';");
      List args = List () << "~/*.doc"
			  << call_function ("GLOB_TILDE");
      List files = call_function ("glob", args, LIST);
      for (size_t i = 0; i < files.size(); i++)
	cerr << (string) files[i] << endl;
    }
}

static List my_cb (List& args, Context cx);
static Scalar my_hashref_cb (Scalar& self, Hashref& args);
static Scalar doit (Scalar& s)
{
  if (double (s) <= 0)
    throw new Exception (string ("'") .append (s)
			 .append ("' is not a positive number"));
  return sqrt (double (s));
}

void
test_cb ()
{
  define_sub ("Foo", "fa", (sub) my_cb);
  cerr << call_function ("Foo::fa", List () << 2 << 3 << 5 << 7, SCALAR);
  call_function ("Foo::fa", List () << 2 << 3 << 5 << 7, VOID);

  Scalar size, ans1, ans2;
  List l (call_function ("Foo::fa", List () << 2 << 3 << 5 << 7, LIST));
  l >> size >> ans1 >> ans2;
  cerr << int (size) << " " << int (ans1) << " " << int (ans2) << endl;

  p->define_sub ("Foo", "fah", my_hashref_cb);
  cerr << p->eval_string ("Foo->fah (baz => 21, foo => 'bla')");

  define_sub ("main", "do_it", doit);
  cerr << eval_string ("do_it(42)") .as_double () << endl;

  eval_string ("eval { do_it(-100); }; warn \"got error: $@\" if $@");
}

static Scalar
my_hashref_cb (Scalar& self, Hashref& args)
{
  cerr << "self: " << string(self) << endl;
  cerr << "foo arg: " << args .fetch ("foo") .as_string () << endl;
  // trigger an uninit warning
  cerr << "bar arg: " << args .fetch ("bar") .as_string () << endl;
  return 2 * (int) args .fetch ("baz");
}

static List my_cb (List& args, Context cx)
{
  if (cx == VOID)
    {
      cerr << "my_cb called in void context." << endl;
      return List ();
    }
  return List () << args .size ()
		 << int (args[0]) + int (args[1])
		 << int (args[2]) + int (args[3]);
}


scalar.cc
/* 
   Copyright (C) 2000 by John Tobey,
   jtobey@john-edwin-tobey.org.  All rights reserved.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; see the file COPYING.  If not, write to the
   Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
   MA 02111-1307  USA
*/

#include "pickle_int.hh"


namespace Pickle
{

  inline static SV*
  string_to_sv (pTHX_ const string& s)
  {
    return newSVpvn (const_cast<char*> (s.data ()), s.size ());
  }

  inline static string
  sv_to_string (pTHX_ SV* sv)
  {
    STRLEN len;
    const char* p;

    p = SvPV (sv, len);
    return string (p, len);
  }

  Scalar::~Scalar ()
  {
    dInterp;
    SvREFCNT_dec (imp);
  }

  static inline SV*
  new_scalar ()
  {
    dTHX;
    return newSVsv (&PL_sv_undef);
  }

  Scalar::Scalar () : imp (new_scalar ()) {}

  Scalar::Scalar (const Scalar& o) : imp (o.imp)
  {
    dTHX;
    SvREFCNT_inc (imp);
  }

  Scalar&
  Scalar::operator= (const Scalar& o)
  {
    dInterp;
    SV* t;

    t = imp;
    imp = SvREFCNT_inc (o.imp);
    SvREFCNT_dec (t);
    return *this;
  }

  // Not-yet-implemented value-to-interpreter mapping.

  const Interpreter*
  Scalar::get_interpreter () const
  {
    return Interpreter::get_current ();
  }

  Interpreter*
  Scalar::get_interpreter ()
  {
    return Interpreter::get_current ();
  }

  bool
  Scalar::is_scalarref () const
  {
    dInterp;
    djSP;  // tryAMAGICunDEREF needs `sp'.
    // tryAMAGICunDEREF needs a var named `sv'.
    SV* sv = const_cast<SV*> (imp);
    if (! SvROK (sv))
      return false;

#ifndef PERL_5005
    // See pp_rv2sv in pp.c .
    tryAMAGICunDEREF(to_sv);
#endif

    // XXX
    switch (SvTYPE (SvRV (sv)))
      {
      case SVt_PVAV:
      case SVt_PVHV:
      case SVt_PVCV:
      case SVt_PVGV:  // In my book, globs are not scalars.
	  return false;
      }
    return true;
  }

  void
  Scalar::check_scalarref () const
  {
    if (! is_scalarref ())
      throw new Exception ("Not a scalar reference");
  }

  bool
  Scalar::is_arrayref () const
  {
    dInterp;
    djSP;  // tryAMAGICunDEREF needs `sp'.
    SV* sv = imp;  // tryAMAGICunDEREF needs a var named `sv'.
    if (! SvROK (sv))
      return false;

#ifndef PERL_5005
    tryAMAGICunDEREF(to_av);  // See pp_rv2av in pp_hot.c .
#endif
    return SvTYPE (SvRV (sv)) == SVt_PVAV;
  }

  void
  Scalar::check_arrayref () const
  {
    if (! is_arrayref ())
      throw new Exception ("Not an array reference");
  }

  bool
  Scalar::is_hashref () const
  {
    dInterp;
    djSP;  // tryAMAGICunDEREF needs `sp'.
    SV* sv = imp;  // tryAMAGICunDEREF needs a var named `sv'.
    if (! SvROK (sv))
      return false;

#ifndef PERL_5005
    tryAMAGICunDEREF(to_hv);  // See pp_rv2hv in pp_hot.c .
#endif
    // XXX Ignoring pseudohashes.
    return SvTYPE (SvRV (sv)) == SVt_PVHV;
  }

  void
  Scalar::check_hashref () const
  {
    if (! is_hashref ())
      throw new Exception ("Not a hash reference");
  }

  bool
  Scalar::is_globref () const
  {
    dInterp;
    djSP;  // tryAMAGICunDEREF needs `sp'.
    // tryAMAGICunDEREF needs a var named `sv'.
    SV* sv = const_cast<SV*> (imp);
    if (! SvROK (sv))
      return false;

#ifndef PERL_5005
    tryAMAGICunDEREF(to_gv);  // See pp_rv2gv in pp.c .
#endif
    // XXX Leaving out stuff about SVt_PVIO.
    return SvTYPE (SvRV (sv)) == SVt_PVGV;
  }

  void
  Scalar::check_globref () const
  {
    if (! is_globref ())
      throw new Exception ("Not a glob reference");
  }

  bool
  Scalar::is_coderef () const
  {
    //dInterp;
    return SvROK (imp) && SvTYPE (SvRV (imp)) == SVt_PVCV;
  }

  void
  Scalar::check_coderef () const
  {
    if (! is_coderef ())
      throw new Exception ("Not a code reference");
  }

  // Conversion from native C++ types to scalar.

  static inline SV* make (unsigned long i) { dTHX; return newSVnv (i); }
  Scalar::Scalar (unsigned long i) : imp (make (i)) {}

  static inline SV* make (long i) { dTHX; return newSViv (i); }
  Scalar::Scalar (long i) : imp (make (i)) {}

  static inline SV* make (unsigned int i) { dTHX; return newSVnv (i); }
  Scalar::Scalar (unsigned int i) : imp (make (i)) {}

  static inline SV* make (int i) { dTHX; return newSViv (i); }
  Scalar::Scalar (int i) : imp (make (i)) {}

  static inline SV* make (unsigned short i) { dTHX; return newSVnv (i); }
  Scalar::Scalar (unsigned short i) : imp (make (i)) {}

  static inline SV* make (short i) { dTHX; return newSViv (i); }
  Scalar::Scalar (short i) : imp (make (i)) {}

  static inline SV* make (unsigned char i) { dTHX; return newSVnv (i); }
  Scalar::Scalar (unsigned char i) : imp (make (i)) {}

  static inline SV* make (signed char i) { dTHX; return newSViv (i); }
  Scalar::Scalar (signed char i) : imp (make (i)) {}

  static inline SV* make (const string& s)
  { dTHX; return string_to_sv (aTHX_ s); }
  Scalar::Scalar (const string& s) : imp (make (s)) {}

  static inline SV* make (const char* s)
  { dTHX; return newSVpv (const_cast<char*> (s), 0); }
  Scalar::Scalar (const char* s) : imp (make (s)) {}

  static inline SV* make (const char* s, unsigned long len)
  { dTHX; return newSVpvn (const_cast<char*> (s), (STRLEN) len); }
  Scalar::Scalar (const char* s, unsigned long len) : imp (make (s, len)) {}

  static inline SV* make (double d) { dTHX; return newSVnv (d); }
  Scalar::Scalar (double d) : imp (make (d)) {}

  static inline SV* make (float d) { dTHX; return newSVnv (d); }
  Scalar::Scalar (float d) : imp (make (d)) {}

  static inline SV* make (bool b) { dTHX; return b ? &PL_sv_yes : &PL_sv_no; }
  Scalar::Scalar (bool b) : imp (make (b)) {}

  bool
  Scalar::defined () const
  {
    dInterp;
    return SvOK (imp);
  }

  Scalar
  Scalar::ref () const
  {
    dInterp;
    if (! SvROK (imp))
      return &PL_sv_no;
    return sv_reftype (SvRV (imp), 1);
  }

  unsigned long
  Scalar::length () const
  {
    dInterp;
    STRLEN len;
    SvPV (imp, len);
    return len;
  }

  bool
  Scalar::eq (const Scalar& that) const
  {
    dInterp;
    return sv_eq (imp, that.imp);
  }

  // Conversions from scalar to basic C++ types.

  unsigned long
  Scalar::as_ulong () const
  {
    dInterp;
    return SvUV (imp);
  }
  long
  Scalar::as_long () const
  {
    dInterp;
    return SvIV (imp);
  }
  unsigned int
  Scalar::as_uint () const
  {
    dInterp;
    return SvUV (imp);
  }
  int
  Scalar::as_int () const
  {
    dInterp;
    return SvIV (imp);
  }
  unsigned short
  Scalar::as_ushort () const
  {
    dInterp;
    return SvUV (imp);
  }
  short
  Scalar::as_short () const
  {
    dInterp;
    return SvIV (imp);
  }
  unsigned char
  Scalar::as_uchar () const
  {
    dInterp;
    return SvUV (imp);
  }
  signed char
  Scalar::as_schar () const
  {
    dInterp;
    return SvIV (imp);
  }
  string
  Scalar::as_string () const
  {
    dInterp;
    return sv_to_string (aTHX_ imp);
  }
  const char*
  Scalar::as_c_str () const
  {
    dInterp;
#ifdef PERL_5005
    STRLEN n_a;
    return SvPV (imp, n_a);
# else // !PERL_5005
    return SvPV_nolen (imp);
#endif
  }
  double
  Scalar::as_double () const
  {
    dInterp;
    return SvNV (imp);
  }
  float
  Scalar::as_float () const
  {
    dInterp;
    return SvNV (imp);
  }
  bool
  Scalar::as_bool () const
  {
    dInterp;
    return SvTRUE (imp);
  }

  Scalar
  Scalar::can (const string& meth) const
  {
    return call_method ("can", List () << meth);
  }
  bool
  Scalar::isa (const string& package) const
  {
    dInterp;
    return sv_derived_from (imp, const_cast<char*> (package.c_str()));
  }

  ostream& operator << (ostream& os, const Scalar& t)
  {
    dTHX;  // XXX
    SV* arg = const_cast<SV*> (t.imp);
    bool has_xml;

    has_xml = Scalar ("XML::Dumper") .can ("new");

    if (!has_xml && !Scalar ("Data::Dumper") .can ("Dumper"))
      {
	const Interpreter* i = t.get_interpreter ();

	try { i ->require_module ("XML::Dumper"); }
	catch (Exception* e) { delete e; }

	has_xml = Scalar ("XML::Dumper")
	  .call_method ("can", List () << "new");

	if (!has_xml)
	  {
	    i ->require_module ("Data::Dumper");
	    if (!Scalar ("Data::Dumper") .call_method ("can", List ()
						       << "Dumper"))
	      throw new Exception ("Loaded Data::Dumper but"
				   " Data::Dumper->Dumper was not defined");
	  }
      }

    return os << (has_xml ? t .as_xml () : t .as_perl ());
  }

  string
  Scalar::as_xml () const
  {
    get_interpreter() ->require_module ("XML::Dumper");
    return Scalar ("XML::Dumper") .call_method ("new")
      .call_method ("pl2xml", List () << *this);
  }

  string
  Scalar::as_perl () const
  {
    const Interpreter* i (get_interpreter ());
    i ->require_module ("Data::Dumper");
    return i ->call_function ("Data::Dumper::Dumper", List () << *this);
  }

  /* Call a method with arguments in scalar context.  */
  Scalar
  Scalar::call_method (const string& meth, const List& args,
		       Context cx) const
  {
    // Avoid perl_call_method because it cannot trap the no-such-method
    // error.  XXX
    {
      dTHX;  // XXX
      if (! perl_get_cv ("Pickle::call_method", 0))
	get_interpreter () ->eval_string
	  ("sub Pickle::call_method {	\
		my ($meth, $obj, $args) = @_;	\
		$obj->$meth (@$args);		\
	    }");
    }
    return get_interpreter () ->call_function
      ("Pickle::call_method",
       List () << meth << *this << (const Arrayref&) args, cx);
  }

  void
  Scalar::bless (const Scalar& pkgname)
  {
    dInterp;
    STRLEN len;
    const char* pname = SvPV (pkgname .imp, len);
    sv_bless (imp, gv_stashpvn (const_cast<char*> (pname), len, 1));
  }

}


Makefile.PL
# Copyright (C) 2000 by John Tobey,
# jtobey@john-edwin-tobey.org.  All rights reserved.
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful, but
#   WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; see the file COPYING.  If not, write to the
#   Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
#   MA 02111-1307  USA
#
# See README about alternative licensing.

use 5.005;
use ExtUtils::MakeMaker;

WriteMakefile (
	       'NAME' => 'pickle',
	       'VERSION' => '0.5.2',
	       'MAN3PODS' => { 'pickle.pod' =>
			       "\$(INST_MAN3DIR)/pickle.\$(MAN3EXT)" },

	       # Don't put \ at end of lines, MakeMaker adds a second \,
	       # which breaks (GNU) Make.
	       'OBJECT' => q/interpreter$(OBJ_EXT) scalar$(OBJ_EXT)
			     scalarref$(OBJ_EXT) arrayref$(OBJ_EXT)
			     hashref$(OBJ_EXT) coderef$(OBJ_EXT)
			     globref$(OBJ_EXT)/,
	      );

package MY;

sub constants {
    my $self = shift;
    my $ret = $self->SUPER::constants;

    if ($self->{'LD'} =~ /^gcc/) {
	warn <<END;
###
### Your LD is $$self{'LD'}.  This had better know about C++ libraries!
### If you get undefined references to a lot of Standard C++ functions,
### try rerunning Makefile.PL with `LD=g++'.
###
END
    }

    # Don't use MakeMaker's default handling of C/C++ files, which is
    # to build an XS module.  We're building ordinary libs instead.
    $ret =~ s/^INST_(STATIC|DYNAMIC|BOOT).*//gm;  # Yuck.
    return $ret;
}

sub test { <<'DONE' }

test_dynamic: test_pickle_d$(EXE_EXT)
	LD_LIBRARY_PATH=. ./test_pickle_d

test_static: test_pickle$(EXE_EXT)
	./test_pickle

test: test_static

test_pickle_d$(EXE_EXT): test_pickle$(OBJ_EXT) libperlint.$(SO) libpickle.$(SO)
	$(LD) -o $@ test_pickle$(OBJ_EXT) -L. -lpickle -lperlint

test_pickle$(EXE_EXT): test_pickle$(OBJ_EXT) $(OBJECT) perlxsi$(OBJ_EXT)
	$(LD) -o $@ $^ $(EMBED_LDOPTS)

test_pickle$(OBJ_EXT): test_pickle.cc pickle.hh
	$(CC) -o $@ -c test_pickle.cc -I .
DONE

sub postamble { <<'DONE' }

clean ::
	$(RM_F) test_pickle$(OBJ_EXT) perlxsi$(OBJ_EXT) perlxsi.c

pure_install :: install_headers install_pickle install_perlint

install_headers: $(LIBHEADERS)
	$(TEST_D) $(PREFIX)/include || mkdir -p $(PREFIX)/include
	$(CP_P) $(LIBHEADERS) $(PREFIX)/include

install_pickle: $(LIBPICKLE)
	$(TEST_D) $(PREFIX)/lib || mkdir -p $(PREFIX)/lib
	$(CP_P) $(LIBPICKLE) $(PREFIX)/lib
	$(RM_F) $(PREFIX)/lib/libpickle.so
	$(LN_S) $(LIBPICKLE) $(PREFIX)/lib/libpickle.so

install_perlint: $(LIBPERLINT)
	$(TEST_D) $(PREFIX)/lib || mkdir -p $(PREFIX)/lib
	$(CP_P) $(LIBPERLINT) $(PREFIX)/lib
	$(RM_F) $(PREFIX)/lib/libperlint.so
	$(LN_S) $(LIBPERLINT) $(PREFIX)/lib/libperlint.so

realclean purge ::
	$(RM_F) test_pickle$(EXE_EXT) test_pickle_d$(EXE_EXT) \
		libpickle.$(SO) $(LIBPICKLE) libperlint.$(SO) \
		$(LIBPERLINT) pickle.$(MAN3EXT) Makefile Makefile.old

pure_all :: libpickle.$(SO) libperlint.$(SO)

LIBPERLINT = libperlint.$(SO).1
LIBPICKLE = libpickle.$(SO).1
LIBHEADERS = pickle.hh
EMBED_LDOPTS = `$(PERL) -MExtUtils::Embed -e ldopts`

libpickle.$(SO): $(LIBPICKLE)
	$(RM_F) $@
	$(LN_S) $< $@

libperlint.$(SO): $(LIBPERLINT)
	$(RM_F) $@
	$(LN_S) $< $@

# "-u" is supposed to mean require the following symbol.
# XXX -Wl, is gcc-specific.  -h is specific to GNU and Solaris.
$(LIBPERLINT): perlxsi$(OBJ_EXT)
	$(LD) -o $@ $(LDDLFLAGS) -u perl_alloc -u boot_DynaLoader \
		$(EMBED_LDOPTS) perlxsi$(OBJ_EXT) -Wl,-h,$@
	$(CHMOD) $(PERM_RWX) $@

perlxsi.c:
	$(PERL) -MExtUtils::Embed -e xsinit

# XXX -Wl, is gcc-specific.  -h is specific to GNU and Solaris.
$(LIBPICKLE): $(OBJECT)
	$(LD) -o $@ $(OBJECT) $(LDDLFLAGS) -Wl,-h,$@
	$(CHMOD) $(PERM_RWX) $@

LN_S = ln -s
CP_P = $(CP) -p
TEST_D = test -d

$(OBJECT): pickle.hh

interpreter$(OBJ_EXT) scalar$(OBJ_EXT) scalarref$(OBJ_EXT) \
	arrayref$(OBJ_EXT) hashref$(OBJ_EXT) coderef$(OBJ_EXT) \
	globref$(OBJ_EXT) : \
	pickle_int.hh

test_pickle$(OBJ_EXT): pickle.hh
DONE
